File: v1/remotix-rust/src/uscita.rs
//! Chi si accorge che la sessione **se ne sta andando**, e lo sa subito.
//!
//! # Il difetto che questo modulo toglie
//!
//! Fino al 3 agosto REMOTIX veniva a sapere di un «Esci» per ultimo: se ne
//! accorgeva quando moriva il flusso di cattura, cioe' quando GNOME aveva
//! **finito** di smontare la sessione. Misurato dal telefono: fra il tocco su
//! «Log Out» e quel momento passavano **5,1 secondi**, e in tutti e cinque il
//! client restava fermo sull'ultimo fotogramma — che, non avendo piu' finestre
//! aperte, era uno sfondo pulito, cioe' *indistinguibile da un desktop vivo*.
//!
//! Da qui il difetto segnalato dall'utente: «la connessione sembra restare
//! viva». Non era un'impressione: era esattamente cio' che il client aveva
//! motivo di credere.
//!
//! # Perche' non basta chiudere e basta
//!
//! Perche' chiudere lo facevamo gia'. IronRDP, quando il flusso di
//! aggiornamenti finisce, arriva a `RunState::Disconnect`, che restituisce il
//! socket e **non manda niente al client**: ne' un codice d'errore, ne' un
//! congedo. Letto sul codice di `ironrdp-server` 0.13, non dedotto. Il client
//! riceve solo una chiusura di socket, e ognuno ne fa quel che vuole: xfreerdp
//! esce, il client Android resta li'.
//!
//! Il congedo dichiarato — `LogoffByUser`, che nel protocollo esiste — richiede
//! un contributo a monte a IronRDP, perche' dall'API pubblica non c'e' modo di
//! spedirlo. Nell'attesa si fa la cosa che **nessun client puo' ignorare**:
//! `SO_LINGER` a zero, cioe' un RST invece di una chiusura educata. Non e' bello
//! quanto un congedo — il client dira' «connessione persa» invece di «sei
//! uscito» — ma arriva mentre l'utente ha ancora il dito sul pulsante che ha
//! appena premuto, e un errore che arriva quando te lo aspetti non e' un guasto:
//! e' una conferma.
//!
//! # Come si sa in tempo
//!
//! Registrandosi con `gnome-session` come fa una qualunque applicazione
//! (`RegisterClient`). Da quel momento il gestore di sessione manda **a noi**,
//! di persona, i tre segnali dell'uscita. Misurato nella VM:
//!
//! | | |
//! |---|---|
//! | logout richiesto | |
//! | `QueryEndSession` | +9 ms — la **domanda**: qualcuno si oppone? |
//! | `EndSession` | +16 ms — la **decisione**: si esce |
//! | `Stop` | +20 ms |
//! | il flusso di cattura muore | +324 ms *(qui ce ne accorgevamo prima)* |
//!
//! Ci si aggancia a `EndSession`, **non** a `QueryEndSession`: la prima e' una
//! decisione, la seconda una domanda a cui qualcuno puo' ancora rispondere di
//! no, e in quel caso GNOME annulla l'uscita. Agganciandosi alla domanda si
//! butterebbe fuori l'utente per un logout che poi non avviene.
//!
//! # La regola dell'ostaggio
//!
//! **Si risponde sempre, e si risponde per primo.** Un client registrato che
//! non riscontra i segnali **blocca l'uscita dell'utente**: gnome-session
//! aspetta lui, e la sessione resta in piedi senza che niente, sullo schermo,
//! spieghi perche'. Non e' un timore: e' successo durante la prova, con una
//! spia che rispondeva alla domanda ma non alla decisione — la sessione e'
//! rimasta in ostaggio finche' la spia non e' scaduta, mezzo minuto dopo.
//!
//! Per questo la risposta parte **prima** di qualsiasi altra cosa, e da un
//! percorso che non puo' fermarsi ad aspettare nient'altro. Tutto il resto —
//! avvisare chi serve la connessione, smontare — viene dopo.
//!
//! # E se gnome-session non ci vuole
//!
//! Si prosegue senza, dichiarandolo nel registro, e si resta con il
//! comportamento di prima: ce ne accorgeremo alla morte del flusso di cattura,
//! cinque secondi dopo. E' peggio, ma e' cio' che c'era, e non e' un motivo per
//! non servire nessuno.

use core::time::Duration;
use std::sync::Arc;

use anyhow::{Context as _, Result};
use futures_util::StreamExt as _;
use tokio::sync::watch;
use tracing::{debug, info, warn};
use zbus::zvariant::OwnedObjectPath;
use zbus::{Connection, Proxy};

use crate::sessione;

/// Ogni quanto si riprova a registrarsi quando non c'e' una sessione.
const ATTESA_SESSIONE: Duration = Duration::from_millis(500);

/// Il nome con cui ci presentiamo a gnome-session.
const NOME: &str = "remotix";

/// Sa dire quando la sessione grafica ha cominciato ad andarsene.
///
/// Chi ascolta non fa I/O: lo stato lo tiene aggiornato un compito a parte.
/// Serve perche' la reazione deve essere **immediata** — e' tutto il punto del
/// modulo — e una chiamata D-Bus nel percorso della reazione rimetterebbe
/// esattamente il ritardo che stiamo togliendo.
pub struct Uscita {
    /// Si incrementa a ogni annuncio. Un contatore e non un booleano perche'
    /// gli annunci sono ripetibili: dopo un logout ne nasce un'altra sessione,
    /// che a sua volta un giorno uscira'.
    annunci: watch::Sender<u64>,
}

impl Uscita {
    /// Comincia a sorvegliare. Non attende: se la sessione non c'e' ancora, la
    /// registrazione avverra' quando ci sara'.
    pub fn avvia() -> Arc<Self> {
        let (annunci, _) = watch::channel(0);
        let uscita = Arc::new(Self { annunci });
        tokio::spawn(sorveglia(Arc::clone(&uscita)));
        uscita
    }

    /// Si risolve al **prossimo** annuncio di uscita.
    ///
    /// «Prossimo» e non «se ce n'e' stato uno»: chi chiama sta servendo una
    /// connessione, e gli annunci delle sessioni precedenti non la riguardano.
    pub async fn attendi_annuncio(&self) {
        let mut ricevitore = self.annunci.subscribe();
        if ricevitore.changed().await.is_err() {
            // Il sorvegliante e' morto: non arrivera' mai nessun annuncio, e
            // restare in attesa per sempre e' la cosa giusta — l'alternativa,
            // risolversi, farebbe cadere la connessione senza motivo.
            core::future::pending::<()>().await;
        }
    }
}

/// Il ciclo che tiene viva la registrazione.
///
/// Si rifa' a ogni sessione: `RegisterClient` vale per la sessione in cui si
/// chiama, e dopo un «Esci» ne nasce una nuova, con un gestore nuovo che di noi
/// non sa niente.
async fn sorveglia(uscita: Arc<Uscita>) {
    loop {
        // Registrarsi senza sessione non ha senso: il gestore non c'e'.
        while !sessione::viva().await {
            tokio::time::sleep(ATTESA_SESSIONE).await;
        }

        match registra().await {
            Ok(cliente) => {
                info!(percorso = %cliente.path(), "registrati con gnome-session: l'uscita si sapra' subito");
                servi_segnali(&cliente, &uscita).await;
                debug!("registrazione finita: la sessione non c'e' piu'");
            }
            Err(errore) => {
                warn!(
                    errore = %format!("{errore:#}"),
                    "registrazione con gnome-session fallita: dell'uscita ci si accorgera' tardi, alla morte della cattura"
                );
                // Non si ritenta a raffica: se il gestore rifiuta, rifiutera'
                // anche fra mezzo secondo, e il registro non deve riempirsi.
                tokio::time::sleep(Duration::from_secs(5)).await;
            }
        }

        // Si aspetta che la sessione se ne vada davvero prima di riprovare,
        // altrimenti ci si riregistrerebbe a raffica su quella morente.
        while sessione::viva().await {
            tokio::time::sleep(ATTESA_SESSIONE).await;
        }
    }
}

/// Si presenta a gnome-session e restituisce l'oggetto che ci rappresenta.
async fn registra() -> Result<Proxy<'static>> {
    let connessione = Connection::session()
        .await
        .context("connessione al bus di sessione")?;

    let gestore = Proxy::new(
        &connessione,
        "org.gnome.SessionManager",
        "/org/gnome/SessionManager",
        "org.gnome.SessionManager",
    )
    .await
    .context("gestore di sessione")?;

    // Il secondo argomento e' l'identificatore di avvio, che si usa per
    // riagganciare un'applicazione ripristinata da una sessione salvata. Non
    // e' il nostro caso: si manda vuoto, come fa qualunque programma avviato a
    // mano.
    let percorso: OwnedObjectPath = gestore
        .call("RegisterClient", &(NOME, ""))
        .await
        .context("RegisterClient")?;

    Proxy::new(
        &connessione,
        "org.gnome.SessionManager",
        percorso,
        "org.gnome.SessionManager.ClientPrivate",
    )
    .await
    .context("oggetto del client di sessione")
}

/// Ascolta i tre segnali finche' la sessione c'e'.
async fn servi_segnali(cliente: &Proxy<'static>, uscita: &Uscita) {
    let domande = cliente.receive_signal("QueryEndSession").await;
    let decisioni = cliente.receive_signal("EndSession").await;
    let arresti = cliente.receive_signal("Stop").await;

    let (Ok(mut domande), Ok(mut decisioni), Ok(mut arresti)) = (domande, decisioni, arresti)
    else {
        warn!("segnali di gnome-session non sottoscritti: dell'uscita ci si accorgera' tardi");
        return;
    };

    loop {
        tokio::select! {
            // La domanda: «qualcuno si oppone?». Si risponde di no e **non si
            // fa altro**. Chi si aggancia qui butta fuori l'utente per un
            // logout che un'applicazione puo' ancora annullare.
            Some(_) = domande.next() => {
                debug!("gnome-session chiede se si puo' uscire");
                riscontra(cliente).await;
            }

            // La decisione. Prima si risponde — la regola dell'ostaggio — e
            // **poi** si avvisa chi sta servendo la connessione.
            Some(_) = decisioni.next() => {
                riscontra(cliente).await;
                info!("la sessione grafica sta uscendo: lo si sa subito, non fra cinque secondi");
                uscita.annunci.send_modify(|n| *n += 1);
            }

            Some(_) = arresti.next() => {
                debug!("gnome-session ci dice di fermarci");
                return;
            }

            else => return,
        }
    }
}

/// Riscontra all'istante, e non lascia che un errore diventi un ostaggio.
///
/// Se la risposta non parte, gnome-session ci aspetta. Non c'e' niente di utile
/// da fare in quel caso se non dirlo forte nel registro: e' il difetto che
/// lascerebbe l'utente senza la possibilita' di sloggarsi, e chi legge deve
/// poterlo riconoscere subito.
async fn riscontra(cliente: &Proxy<'static>) {
    match cliente
        .call::<_, _, ()>("EndSessionResponse", &(true, ""))
        .await
    {
        Ok(()) => debug!("riscontrato a gnome-session: per noi si puo' uscire"),
        Err(errore) => warn!(
            errore = %errore,
            "riscontro a gnome-session fallito: l'uscita dell'utente potrebbe restare bloccata"
        ),
    }
}

File: fondamenta/remotix-c/src/uscita.c
#include "uscita.h"

#include <gio/gio.h>

#include "registro.h"
#include "sessione.h"

#define NOME_GESTORE "org.gnome.SessionManager"
#define PERCORSO_GESTORE "/org/gnome/SessionManager"
#define IFACE_GESTORE "org.gnome.SessionManager"
#define IFACE_CLIENTE "org.gnome.SessionManager.ClientPrivate"

/* Il nome con cui ci presentiamo a gnome-session. */
#define NOME_CLIENTE "remotix"

/* Ogni quanto si riprova a registrarsi, e ogni quanto si controlla che la
 * sessione a cui siamo registrati esista ancora. */
#define CADENZA_MS 500
#define GIRI_PER_CONTROLLO 10 /* cioe' ogni 5 secondi */

struct Uscita
{
	GThread *thread;
	GMainContext *contesto;
	GMainLoop *ciclo;

	UscitaAnnuncio su_uscita;
	gpointer dati;

	GDBusConnection *bus;
	char *cliente; /* percorso dell'oggetto che ci rappresenta */
	guint sottoscrizioni[3];
	guint giri;
	/* Il guasto della registrazione si dice una volta: il battito riprova due
	 * volte al secondo, e su un compositore che non e' GNOME non riuscira' mai. */
	gboolean detto_il_guasto;

	TipoCompositore tipo;
	/* Su KDE non ci si registra: si GUARDA.  Vedi `sorveglia_kde`. */
	guint spia_kde;
	gboolean uscita_annunciata;
};

/*
 * Riscontra all'istante, e non aspetta la risposta.
 *
 * ⛔ Asincrona di proposito: questo e' il percorso su cui pende l'uscita
 *    dell'utente, e una chiamata che attende puo' trattenerlo.  Se il riscontro
 *    non parte non c'e' niente di utile da fare se non dirlo FORTE nel
 *    registro: e' il difetto che lascerebbe l'utente senza la possibilita' di
 *    sloggarsi, e chi legge deve poterlo riconoscere subito.
 */
static void su_riscontro(GObject *sorgente, GAsyncResult *esito, gpointer dati)
{
	g_autoptr(GError) sbaglio = NULL;
	g_autoptr(GVariant) risposta =
	    g_dbus_connection_call_finish(G_DBUS_CONNECTION(sorgente), esito, &sbaglio);

	if (!risposta)
		avviso("riscontro a gnome-session fallito (%s): l'uscita dell'utente potrebbe "
		       "restare BLOCCATA",
		       sbaglio->message);
	else
		traccia("riscontrato a gnome-session: per noi si puo' uscire");
}

static void riscontra(Uscita *uscita)
{
	if (!uscita->cliente)
		return;
	g_dbus_connection_call(uscita->bus, NOME_GESTORE, uscita->cliente, IFACE_CLIENTE,
	                       "EndSessionResponse", g_variant_new("(bs)", TRUE, ""), NULL,
	                       G_DBUS_CALL_FLAGS_NONE, 2000, NULL, su_riscontro, NULL);
}

/*
 * Ci si toglie dall'elenco dei client di gnome-session.
 *
 * ⛔ Registrarsi ha un prezzo che va pagato per intero: da quel momento il
 *    gestore di sessione ci considera **suoi**, e alla fine della sessione fa
 *    con noi quel che fa con le applicazioni — ci dice `Stop` e si aspetta che
 *    ce ne andiamo.  Ma REMOTIX non deve andarsene: deve restare in ascolto e
 *    riavviare la sessione alla connessione successiva (§5.9-bis di
 *    SPECIFICA.md), altrimenti dopo un «Esci» l'utente si ritrova un server
 *    morto e nessun desktop.
 *
 *    Quindi appena si sa che la sessione sta finendo ci si sfila.
 */
static void disiscrivi(Uscita *uscita)
{
	if (!uscita->bus || !uscita->cliente)
		return;
	g_dbus_connection_call(uscita->bus, NOME_GESTORE, PERCORSO_GESTORE, IFACE_GESTORE,
	                       "UnregisterClient", g_variant_new("(o)", uscita->cliente), NULL,
	                       G_DBUS_CALL_FLAGS_NONE, 2000, NULL, NULL, NULL);
	diagnostica("disiscritti da gnome-session: la nostra vita torna nostra");
}

static void sgancia(Uscita *uscita)
{
	disiscrivi(uscita);
	for (gsize i = 0; i < G_N_ELEMENTS(uscita->sottoscrizioni); i++)
	{
		if (uscita->sottoscrizioni[i])
			g_dbus_connection_signal_unsubscribe(uscita->bus, uscita->sottoscrizioni[i]);
		uscita->sottoscrizioni[i] = 0;
	}
	g_clear_pointer(&uscita->cliente, g_free);
}

static void su_segnale(GDBusConnection *bus, const char *mittente, const char *percorso,
                       const char *interfaccia, const char *segnale, GVariant *parametri,
                       gpointer dati)
{
	Uscita *uscita = dati;

	if (g_strcmp0(segnale, "QueryEndSession") == 0)
	{
		/* La DOMANDA.  Si risponde di no-non-mi-oppongo e non si fa altro: chi
		 * si aggancia qui butta fuori l'utente per un logout che
		 * un'applicazione puo' ancora annullare. */
		diagnostica("gnome-session chiede se si puo' uscire");
		riscontra(uscita);
		return;
	}

	if (g_strcmp0(segnale, "EndSession") == 0)
	{
		/* La DECISIONE.  Prima si risponde — la regola dell'ostaggio — e POI si
		 * avvisa chi sta servendo la connessione. */
		riscontra(uscita);
		informazione("la sessione grafica sta uscendo: lo si sa subito, non fra cinque secondi");
		if (uscita->su_uscita)
			uscita->su_uscita(uscita->dati);
		/* E ci si sfila SUBITO: da qui in poi gnome-session non deve piu'
		 * considerarci uno dei suoi.  Il `battito` ci riregistrera' quando
		 * nascera' la sessione successiva. */
		sgancia(uscita);
		return;
	}

	if (g_strcmp0(segnale, "Stop") == 0)
	{
		diagnostica("gnome-session ci dice di fermarci");
		sgancia(uscita);
	}
}

static void sottoscrivi(Uscita *uscita)
{
	static const char *segnali[] = { "QueryEndSession", "EndSession", "Stop" };

	for (gsize i = 0; i < G_N_ELEMENTS(segnali); i++)
		uscita->sottoscrizioni[i] = g_dbus_connection_signal_subscribe(
		    uscita->bus, NOME_GESTORE, IFACE_CLIENTE, segnali[i], uscita->cliente, NULL,
		    G_DBUS_SIGNAL_FLAGS_NONE, su_segnale, uscita, NULL);
}

static gboolean registra(Uscita *uscita)
{
	g_autoptr(GError) sbaglio = NULL;
	g_autoptr(GVariant) risposta = NULL;

	/* Dopo un logout il bus dell'utente e' un altro: quello che tenevamo qui e'
	 * chiuso, e riregistrarsi su una connessione chiusa non fallisce con un
	 * errore — non fa proprio niente. */
	if (uscita->bus && g_dbus_connection_is_closed(uscita->bus))
		g_clear_object(&uscita->bus);

	if (!uscita->bus)
	{
		uscita->bus = sessione_bus(&sbaglio);   /* mai g_bus_get_sync: vedi sessione.h */
		if (!uscita->bus)
			return FALSE;
	}

	/* Il secondo argomento e' l'identificatore di avvio, che serve a riagganciare
	 * un'applicazione ripristinata da una sessione salvata.  Non e' il nostro
	 * caso: si manda vuoto, come qualunque programma avviato a mano. */
	risposta = g_dbus_connection_call_sync(
	    uscita->bus, NOME_GESTORE, PERCORSO_GESTORE, IFACE_GESTORE, "RegisterClient",
	    g_variant_new("(ss)", NOME_CLIENTE, ""), G_VARIANT_TYPE("(o)"), G_DBUS_CALL_FLAGS_NONE,
	    5000, NULL, &sbaglio);
	if (!risposta)
	{
		/*
		 * ⛔ SI DICE UNA VOLTA SOLA, e non e' pignoleria di stile.
		 *
		 *    Il battito riprova ogni mezzo secondo finche' una sessione c'e': su
		 *    un compositore che non e' GNOME `org.gnome.SessionManager` non
		 *    esistera' mai, e questa riga usciva DUE VOLTE AL SECONDO per tutta la
		 *    durata del servizio.  Un registro cosi' non e' rumoroso: e' inservibile
		 *    — la riga che spiega il guasto vero ci finisce sepolta dentro, ed e'
		 *    esattamente quel che e' successo alla prima prova su KDE, dove il
		 *    motivo per cui il server moriva stava fra due schermate di questa.
		 *
		 * ⏸ Su KDE la sentinella dell'uscita e' un'altra: due sottoscrizioni
		 *    passive a `org.kde.Shutdown` e `org.kde.KWinWrapper` (`kde.md` §6.5),
		 *    che non mettono in gioco la sessione dell'utente.  E' la VOCE 3 del
		 *    piano di fase 11.
		 */
		if (!uscita->detto_il_guasto)
		{
			uscita->detto_il_guasto = TRUE;
			avviso("registrazione con gnome-session fallita (%s): dell'uscita ci si accorgera' "
			       "tardi, alla morte della cattura — e non lo ripeto piu'",
			       sbaglio->message);
		}
		return FALSE;
	}

	g_variant_get(risposta, "(o)", &uscita->cliente);
	sottoscrivi(uscita);
	informazione("registrati con gnome-session (%s): l'uscita si sapra' subito", uscita->cliente);
	return TRUE;
}

/*
 * Il battito che tiene viva la registrazione.
 *
 * `RegisterClient` vale per la sessione in cui lo si chiama: dopo un «Esci» ne
 * nasce una nuova, con un gestore nuovo che di noi non sa niente.  Quindi ci si
 * riregistra, e lo si fa solo quando una sessione c'e' davvero.
 */
/*
 * La sentinella dell'uscita su KDE, e perche' e' PASSIVA.
 *
 * ⛔ SU KDE NON ESISTE UN `RegisterClient`, e la cosa che gli somiglia e'
 *    peggio del problema.  L'equivalente vero e' XSMP su ICE — serve `libSM`,
 *    `libICE` e un `DISPLAY` — e `org.kde.KSMServerInterface` non ha alcun
 *    segnale «la sessione sta finendo» [✗].  Chi si registrasse presso ksmserver
 *    erediterebbe la regola dell'ostaggio nella sua forma peggiore: un client
 *    che non risponde FRENA IL LOGOUT DELL'UTENTE DI QUINDICI SECONDI
 *    (`ksmserver/logout.cpp:293-303`).
 *
 *    Si guarda invece un nome sul bus: `org.kde.Shutdown` e' **attivabile**, cioe'
 *    non esiste finche' il logout non comincia — e compare esattamente allora
 *    (`plasma-shutdown/shutdown.cpp:20-23`).  E' la stessa cosa che fa
 *    `startplasma` per sapere di dover uscire (`startplasma.cpp:673-689`).
 *    Costa una sottoscrizione e non mette in gioco la sessione dell'utente.
 *
 * ⚠ Con una differenza rispetto a GNOME che va detta, perche' cambia i tempi:
 *   li' `EndSession` e' una DECISIONE che si riscontra, e riscontrarla per primi
 *   accorcia l'uscita.  Qui non c'e' niente da riscontrare: si sa che sta
 *   succedendo, e basta.  Il client cade lo stesso subito, perche' quel che
 *   conta e' saperlo presto — ed e' quel che questa spia da'.
 */
static void su_nome_cambiato(GDBusConnection *bus, const char *mittente, const char *percorso,
                             const char *interfaccia, const char *segnale, GVariant *parametri,
                             gpointer dati)
{
	Uscita *uscita = dati;
	const char *nome = NULL, *prima = NULL, *dopo = NULL;

	g_variant_get(parametri, "(&s&s&s)", &nome, &prima, &dopo);
	/* Compare un proprietario: il logout e' cominciato ADESSO. */
	if (!dopo || !*dopo || uscita->uscita_annunciata)
		return;

	uscita->uscita_annunciata = TRUE;
	informazione("la sessione di KDE sta uscendo: lo si sa subito, non alla morte della cattura");
	if (uscita->su_uscita)
		uscita->su_uscita(uscita->dati);
}

static void sorveglia_kde(Uscita *uscita)
{
	g_autoptr(GError) sbaglio = NULL;

	if (uscita->spia_kde)
		return;
	if (!uscita->bus)
	{
		uscita->bus = sessione_bus(&sbaglio);
		if (!uscita->bus)
		{
			if (!uscita->detto_il_guasto)
			{
				uscita->detto_il_guasto = TRUE;
				avviso("nessun bus di sessione (%s): dell'uscita ci si accorgera' tardi",
				       sbaglio->message);
			}
			return;
		}
	}

	uscita->spia_kde = g_dbus_connection_signal_subscribe(
	    uscita->bus, "org.freedesktop.DBus", "org.freedesktop.DBus", "NameOwnerChanged",
	    "/org/freedesktop/DBus", "org.kde.Shutdown", G_DBUS_SIGNAL_FLAGS_NONE, su_nome_cambiato,
	    uscita, NULL);
	informazione("sentinella dell'uscita: sorveglio org.kde.Shutdown");
}

static gboolean battito(gpointer dati)
{
	Uscita *uscita = dati;

	if (uscita->tipo == COMPOSITORE_KWIN)
	{
		sorveglia_kde(uscita);
		/*
		 * L'annuncio si riarma quando la sessione se n'e' andata davvero: dopo un
		 * «Esci» ne nascera' un'altra, e di quella si dovra' sapere l'uscita
		 * altrettanto presto.
		 */
		if (uscita->uscita_annunciata && !sessione_viva())
			uscita->uscita_annunciata = FALSE;
		return G_SOURCE_CONTINUE;
	}

	if (!uscita->cliente)
	{
		if (sessione_viva())
			registra(uscita);
		return G_SOURCE_CONTINUE;
	}

	/* Registrati: i segnali fanno il lavoro.  Si controlla comunque ogni tanto
	 * che la sessione ci sia ancora, perche' una sessione che muore male non
	 * manda alcuno `Stop` e resteremmo agganciati a un gestore che non c'e'. */
	if (++uscita->giri >= GIRI_PER_CONTROLLO)
	{
		uscita->giri = 0;
		if (!sessione_viva())
		{
			diagnostica("la sessione a cui eravamo registrati non c'e' piu'");
			sgancia(uscita);
		}
	}
	return G_SOURCE_CONTINUE;
}

static gpointer thread_uscita(gpointer dati)
{
	Uscita *uscita = dati;
	GSource *tic;

	g_main_context_push_thread_default(uscita->contesto);

	tic = g_timeout_source_new(CADENZA_MS);
	g_source_set_callback(tic, battito, uscita, NULL);
	g_source_attach(tic, uscita->contesto);

	g_main_loop_run(uscita->ciclo);

	g_source_destroy(tic);
	g_source_unref(tic);
	if (uscita->spia_kde && uscita->bus)
	{
		g_dbus_connection_signal_unsubscribe(uscita->bus, uscita->spia_kde);
		uscita->spia_kde = 0;
	}
	sgancia(uscita);
	g_main_context_pop_thread_default(uscita->contesto);
	return NULL;
}

Uscita *uscita_avvia(UscitaAnnuncio su_uscita, gpointer dati, TipoCompositore tipo)
{
	Uscita *uscita = g_new0(Uscita, 1);

	uscita->su_uscita = su_uscita;
	uscita->dati = dati;
	uscita->tipo = tipo;
	uscita->contesto = g_main_context_new();
	uscita->ciclo = g_main_loop_new(uscita->contesto, FALSE);
	uscita->thread = g_thread_new("remotix-uscita", thread_uscita, uscita);
	return uscita;
}

void uscita_ferma(Uscita *uscita)
{
	if (!uscita)
		return;
	if (uscita->ciclo)
		g_main_loop_quit(uscita->ciclo);
	if (uscita->thread)
		g_thread_join(uscita->thread);
	g_clear_pointer(&uscita->ciclo, g_main_loop_unref);
	g_clear_pointer(&uscita->contesto, g_main_context_unref);
	g_clear_object(&uscita->bus);
	g_free(uscita->cliente);
	g_free(uscita);
}
